### PR TITLE
un-indenting head-safe check to apply for cached files too

### PR DESCRIPTION
**Pull request by eas342**
Un-indenting the head-safe boolean check to apply to both fixing incomplete cached files and new files. This appears to fix this issue:
https://github.com/astropy/astroquery/issues/3231

However, I do not understand what this `head_safe` keyword does, so please check that this does not break the intended functionality or flow.

### CODE REVIEW

**Review by snbianco — COMMENTED** [review]
See other comment!

**Review by keflavich — CHANGES_REQUESTED** [review]
Unindenting this block is not the correct behavior; it will overwrite the `continuation` block above, which would result in the download continuation failing - it will download the whole file and append it to whatever was partly completed.

However, this same code block (the `if head_safe: response = request(...)` block) needs to be repeated above, in the other location `open_mode='wb'`

**Comment by eas342** [comment]
Thank you @keflavich ,

Unfortunately, the change you made does not seem to pass https://github.com/astropy/astroquery/issues/3231

import os

from astroquery.mast import Observations
from astropy.io import fits

    import os
    from astropy.io import fits
    from astroquery.mast import Observations
    fileN = 'jw01185103001_02102_00001-seg001_nrcalong_rate.fits'
    #os.remove(fileN)
    Observations.download_file('mast:jwst/product/' + fileN)
    with open(fileN, 'r+') as f:
        f.seek(1000)
        f.truncate()
    Observations.download_file('mast:jwst/product/' + fileN)
    tmpHDU = fits.open(fileN)

WARNING: Found cached file jw01185103001_02102_00001-seg001_nrcalong_rate.fits with size 0 that is different from expected size 83520 [astroquery.query]
Downloading URL https://mast.stsci.edu/api/v0.1/Download/file?uri=mast:jwst/product/jw01185103001_02102_00001-seg001_nrcalong_rate.fits to jw01185103001_02102_00001-seg001_nrcalong_rate.fits ...
|==========================================|  83k/ 83k (100.00%)         0s
WARNING: Found cached file jw01185103001_02102_00001-seg001_nrcalong_rate.fits with size 1000 that is different from expected size 83520 [astroquery.query]
Downloading URL https://mast.stsci.edu/api/v0.1/Download/file?uri=mast:jwst/product/jw01185103001_02102_00001-seg001_nrcalong_rate.fits to jw01185103001_02102_00001-seg001_nrcalong_rate.fits ...
|==========================================|  83k/ 83k (100.00%)         0s
OSError: Empty or corrupt FITS file

**Comment by keflavich** [comment]
Thanks @eas342.  Indeed, there was another corner case we hadn't checked.  The latest commit catches it.

**Comment by eas342** [comment]
That works, super! Thank you @keflavich

**Comment by bsipocz** [comment]
@eas342 @keflavich - I'm not sure how to test this in CI, etc. But if you are certain it covers all the cases, and add a changelog, it can still go in the next release ~tomorrow.
But if there is no rush or there is any uncertainty, then I could sit the release out and get it merged after it's out to have it in `main` for a little longer.

**Comment by keflavich** [comment]
let's add tests; i think i can turn @eas342 's mwe into a test easily enough.  so delay merging

File: astroquery/query.py [hunk 1]
```diff
@@ -466,17 +472,26 @@ def _download_file(self, url, local_filepath, timeout=None, auth=None,
...
-                             .format(local_filepath, statinfo.st_size))
+                    log.info(f"Found cached file {local_filepath} with expected size {statinfo.st_size}.")
                     response.close()
-                    return
+                    return local_filepath
             else:
-                log.info("Found cached file {0}.".format(local_filepath))
+                # This case doesn't appear reachable under normal circumstances
+                # It is not covered by tests, and probably indicates a badly-behaved server
+                log.info(f"Found cached file {local_filepath}.  Could not verify length.")
```
Comment (bsipocz) [inline astroquery/query.py:492]: we may then want to raise an exception for these and even delete the cached file?

File: astroquery/tests/test_query.py [hunk 2]
```diff
@@ -1,39 +1,388 @@
...
+from requests.structures import CaseInsensitiveDict
 from astroquery.query import BaseQuery, BaseVOQuery
+from astroquery.utils.mocks import MockResponse
+from itertools import product
 
+# Test data directory
+DATA_DIR = Path(__file__).parent / 'data'
+DATA_DIR.mkdir(exist_ok=True)
 
-class with_VO(BaseVOQuery, BaseQuery):
```
Comment (bsipocz) [inline astroquery/tests/test_query.py:5]: Why are these deleted? I want to keep the multiple inheritances case covered as that is what we actually do in the code at a lot of places!

File: CHANGES.rst [hunk 3]
```diff
@@ -8,9 +8,14 @@ New Tools and Services
 API changes
 -----------
 
+query.py
```
Comment (bsipocz) [inline CHANGES.rst:11]: please move this into `Infrastructure, Utility and Other Changes and Additions` and no need to add `query.py` heading.

**Review by bsipocz — COMMENTED** [review]
Some minor comments. 
Please add back the multiple inheritances in the tests, and do a rebase for the changelog. 

If you can squash out/consolidate a few of the flake8 and similar converging commits that would be good, too, but not a big deal if not.

File: astroquery/query.py (hunk 1 above)
Comment (keflavich) [reply astroquery/query.py:492]: that's not a bad idea, I'll do that.

File: astroquery/tests/test_query.py (hunk 2 above)
Comment (keflavich) [reply astroquery/tests/test_query.py:5]: there was no good reason for that.  I used AI to help me write tests, I didn't notice that it was deleting them.  I've added them back.  Thanks for noting this!

File: CHANGES.rst [hunk 4]
```diff
@@ -9,6 +9,11 @@ New Tools and Services
...
+^^^^^^^^
+
+- ``_download_file`` now returns the local file path in all cases (formerly, it would return ``None`` if the file was already cached).  Some corner cases where downloads were not properly continued have been fixed [#3232]
```
Comment (bsipocz) [inline CHANGES.rst:15]: Something went wrong here

**Review by bsipocz — APPROVED** [review]
Looks good, and I'll fixup the changelog prior merging.

File: CHANGES.rst (hunk 4 above)
Comment (keflavich) [reply CHANGES.rst]: 🙄 yep bad squash.

Comment (bsipocz) [reply CHANGES.rst]: still bad (in wrong version section), but I'm pushing the changes now.

**Comment by bsipocz** [comment]
OK, let's merge this now. I didn't see any problems trying it locally, but nevertheless it will be good to have it in for as long as possible before the next release.

File: CHANGES.rst (hunk 4 above)
Comment (keflavich) [reply CHANGES.rst]: ach, sorry, that was just me that time - I didn't realize the existing Infrastructure label was already under a release section

File: astroquery/query.py [hunk 5]
```diff
@@ -466,17 +472,24 @@ def _download_file(self, url, local_filepath, timeout=None, auth=None,
...
                     response.close()
-                    return
+                    return local_filepath
             else:
-                log.info("Found cached file {0}.".format(local_filepath))
-                response.close()
-                return
+                # This case doesn't appear reachable under normal circumstances
+                # It is not covered by tests, and probably indicates a badly-behaved server
+                raise ValueError(f"Found cached file {local_filepath}.  Could not verify length.")
```
Comment (snbianco) [inline astroquery/query.py:492]: This block is causing some issues for our internal MAST test suite. We actually do have certain cases where the response returned by the download request doesn't contain a header for "Content-Length". I'm working to understand these edge cases better, but it may be that there's no way to get the content length, in which case trying to download a file that's already cached fails. This may have been why the `continuation` parameter is set to `False` in MAST download calls.

Would the best course of action be to re-download completely if no content length is found?

Comment (keflavich) [reply astroquery/query.py:492]: @snbianco can you tell us the answer to that question?  What does it mean if there's no content-length?  I'd argue that continuation in that scenario is probably not possible, but I don't understand the case here - why are we downloading a file without a specified length?

iirc, I couldn't test this because httpbin wouldn't serve anything without that header keyword

Comment (snbianco) [reply astroquery/query.py:492]: To the best of my knowledge, there are certain files that are generated and delivered by a different server than the MAST portal. For example, the catalog files for the Hubble Legacy Archive are generated by https://hla.stsci.edu/. So, content length may be unknown to the MAST portal, and the portal doesn't currently return it in the download response.

If it does turn out that we can't get the length of these files from the Portal, redownloading seems like the only way to go. The downside to that is the time lost redownloading when a complete cached file could have been used. Would it be possible to raise a warning instead of a `ValueError` when the content length header is not found? Or add a parameter to control whether a warning or an error is raised?

Comment (snbianco) [reply astroquery/query.py:492]: I've been talking about this issue with a few folks at MAST, and I know a bit more about why these particular files don't have a content length. They are generated on-the-fly by a service that queries our database and puts the results into the requested format. They are also streamed and not staged at all. There is a content length coming from the service, but I assume its being dropped somewhere between there and the portal API. There are a few people looking into it, and I'm hopeful that we can add the header to the files that come from that particular service.

That being said, we might have this issue with other files that are generated dynamically and then streamed from the database. Since this only applies in the case that the file is already cached and continuation is not possible (MAST sets `continuation` to `False` anyway), maybe the default behavior should be to log a warning and re-download the file entirely?
